Denormalize arrays, ids, and arrays of ids

jcarbo opened the pull request:
Fixes #3 

This PR updates the denormalize function to handle denormalizing lists when the schema is an `IterableSchema` (i.e. `arrayOf(someEntitySchema)`). I did some refactoring as part of that work, and a nice side effect was being able to denormalize just by passing the id (or array of ids). This was already how denormalizing nested arrays was handled, so I think it makes sense to expose at the top level.

A cool side effect is the greater symmetry with normalizr. This is now true:

```js
const { result, entities } = normalize(response, schema);
denormalize(result, entities, schema) == response
```

gpbl commented on test/index.js:
```diff
@@ -7,9 +7,9 @@ import { normalize, Schema, arrayOf, unionOf } from 'normalizr';
-  });
+  // it("should return undefined when denormalizing an undefined entity", () => {
+  //   expect(denormalize(undefined)).to.be.undefined;
+  // });
```
Hi @jcarbo thanks, looks like a great improvement. I'll review the PR tomorrow (it's nighttime here).
Is there a reason why this test has been commented out?

jcarbo replied:
Whoops, I had commented out some passing specs to more easily see console logs in failing ones. Looks like I had forgotten to uncomment this one. Fixed!

jcarbo commented:
Just squashed the commits and amended the last one. Initially the updates here allowed the normalized entities (during non-immutable usage) to be mutated. However, I realized this probably isn't ideal especially for people using this with redux since denormalized entities would wind up back in the store.

I added some specs failing specs to confirm and updated the functionality so they pass. It will now safely denormalize without mutating the passed entities.

If all goes well and this winds up getting merged, could you publish a new version to NPM? Thanks!

jcarbo commented:
@gpbl - do you think you'll have a chance to review this over the weekend? I'm currently hard-coding this in a project of mine and would love avoid that 😄

igorbarbashin commented:
I'm voting for this feature! Thanks @jcarbo for implementing

gpbl commented:
Published as 0.3.0 (with the immutable-js support as well). Thanks @jcarbo again 👍 
please forgive me for the delay on merging it 🙏  I hope it didn't cause too much problems with your project!
